File: src/capability/capability_audiovolume.c
/*
 * Copyright (c) 2018 Samsung Electronics Co., Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an AS IS BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "st_things.h"
#include "log.h"

#define VALUE_STR_LEN_MAX 32

static const char* KEY_VOLUME = "volume";
static int64_t g_volume;

static const char* KEY_MUTE = "mute";
static bool g_mute;

extern bool user_player_get_volume(float *vol);
extern bool user_player_is_muted(bool *muted);
extern bool user_player_set_mute(bool muted);
extern bool user_player_set_volume(float vol);

bool handle_get_request_on_resource_capability_audiovolume(st_things_get_request_message_s* req_msg, st_things_representation_s* resp_rep)
{
    DBG("Received a GET request on %s\n", req_msg->resource_uri);
	DBG("current g_volume: [%lld], g_mute:[%d]", g_volume, g_mute);

	if (req_msg->has_property_key(req_msg, KEY_VOLUME)) {
		DBG("key[%s] exists", KEY_VOLUME);
		float temp_vol = 0.0;
		user_player_get_volume(&temp_vol);
		g_volume = (int64_t) (temp_vol * 100);
		DBG("current g_volume: [%lld]", g_volume);
		resp_rep->set_int_value(resp_rep, KEY_VOLUME, g_volume);
	}

	if (req_msg->has_property_key(req_msg, KEY_MUTE)) {
		DBG("key[%s] exists", KEY_MUTE);
		user_player_is_muted(&g_mute);
		DBG("current g_mute: [%d]", g_mute);
		resp_rep->set_bool_value(resp_rep, KEY_MUTE, g_mute);
	}

    return true;
}

bool handle_set_request_on_resource_capability_audiovolume(st_things_set_request_message_s* req_msg, st_things_representation_s* resp_rep)
{
    DBG("Received a SET request on %s\n", req_msg->resource_uri);
	DBG("current g_volume: [%lld], g_mute:[%d]", g_volume, g_mute);

	int64_t ivalue;
	if (req_msg->rep->get_int_value(req_msg->rep, KEY_VOLUME, &ivalue)) {
		DBG("requested volume: [%lld]", ivalue);
		g_volume = ivalue;
		float temp_vol = (float) (g_volume/100.);
		user_player_set_volume(temp_vol);
		resp_rep->set_int_value(resp_rep, KEY_VOLUME, g_volume);
	}

	bool bvalue;
	if (req_msg->rep->get_bool_value(req_msg->rep, KEY_MUTE, &bvalue)) {
		DBG("requested mute: [%d]", bvalue);
		g_mute = bvalue;
		user_player_set_mute(g_mute);
		resp_rep->set_bool_value(resp_rep, KEY_MUTE, g_mute);
	}

	st_things_notify_observers(req_msg->resource_uri);

    return true;
}

File: plugin/plugin/js/capability_audioVolume.js
/*
 * Copyright (c) 2015 - 2017 Samsung Electronics Co., Ltd All Rights Reserved
 *
 * Licensed under the Apache License, Version 2.0 (the License);
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an AS IS BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

var capabilityAudioVolume = {
	'href' : "/capability/audioVolume/main/0",
	'value' : "50",
	'muteState' : false,

	'update' : function() {
		ocfDevice.getRemoteRepresentation(this.href, this.onRepresentCallback);
	},

	'onRepresentCallback' : function(result, deviceHandle, uri, rcsJsonString) {
		scplugin.log.debug(className, arguments.callee.name, result);
		scplugin.log.debug(className, arguments.callee.name, uri);

		if (result == "OCF_OK" || result == "OCF_RESOURCE_CHANGED" || result == "OCF_RES_ALREADY_SUBSCRIBED") {
			if(rcsJsonString["volume"] === undefined || rcsJsonString["mute"] === undefined) return;
			capabilityAudioVolume.value = rcsJsonString["volume"];
			capabilityAudioVolume.muteState = rcsJsonString["mute"];

			var styleText = '#' + 'volume_value' + '::-webkit-slider-runnable-track{background-size:' + capabilityAudioVolume.value + '% 100%} ';
			inlineStyle.textContent = styleText;
			document.getElementById("volume_value").value = capabilityAudioVolume.value;
			var image = document.getElementById("muteButton");
			var volume = document.getElementById("volume_value");
			if (capabilityAudioVolume.muteState == false) {
				image.src = "res/controller_ic_unmute.png";
				volume.disabled = false;
			} else {
				image.src = "res/controller_ic_mute.png";
				volume.disabled = true;
			}
		}
	},

	"clearBg" : function (image) {
		image.style.backgroundColor = "white";
	},

	'setVolume' : function(value) {
		scplugin.log.debug(className, arguments.callee.name, "audioVolume : " + value);
		var setRcsJson = {};
		setRcsJson["volume"] = value;
		setRcsJson["mute"] = this.muteState;
		ocfDevice.setRemoteRepresentation(this.href, setRcsJson, this.onRepresentCallback);
	},

	'setMute' : function(value) {
		scplugin.log.debug(className, arguments.callee.name, "audioVolume (mute): " + value);
		var image = document.getElementById("muteButton");
		image.style.backgroundColor = "cyan";
		var setRcsJson = {};
		setRcsJson["volume"] = this.value;
		setRcsJson["mute"] = value;
		ocfDevice.setRemoteRepresentation(this.href, setRcsJson, this.onRepresentCallback);
		setTimeout(this.clearBg, 500, image);
	},

	'muteToggle' : function() {
		if (this.muteState == false) {
			this.setMute(true);
		} else {
			this.setMute(false);
		}
	}
}
